[Login-UI] Password Strength Indicator

andreas-blaettlinger opened the pull request:
Closes #23573

Very first draft implemented for the register page only. Want to collect some feedback or other proposals.

andreas-blaettlinger committed:
Login-UI Password Strength Indicator

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -104,6 +104,18 @@
                         </#if>
                     </div>
+                    <div class="${properties.kcInputWrapperClass!}" id="password-entropy" data-password-input="password" style="display: none;">
+                        <div class="progress" style="margin-bottom: 0; margin-top: 1px;">
```
If possible, I'd like to avoid in-line styling. Preferably this style would be added to a separate style sheet.

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -104,6 +104,18 @@
                     </div>
+                    <div class="${properties.kcInputWrapperClass!}" id="password-entropy" data-password-input="password" style="display: none;">
+                        <div class="progress" style="margin-bottom: 0; margin-top: 1px;">
+                            <div class="progress-bar" role="progressbar" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">
```
Nice, thanks for taking accessibility into account. I am not quite sure that a progress bar is the correct role though, as it is more intended for reporting progress on long-running tasks.

Perhaps a more suitable role for this one would be the [`status` role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/status_role), which is a live region that uses `polite`, and thus does not interrupt the user.

```suggestion
                            <div class="progress-bar" role="status">
```

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -104,6 +104,18 @@
+                    <div class="${properties.kcInputWrapperClass!}" id="password-entropy" data-password-input="password" style="display: none;">
+                        <div class="progress" style="margin-bottom: 0; margin-top: 1px;">
+                            <div class="progress-bar" role="progressbar" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">
+                                <span
```
Do we need this span element? Perhaps this could all be on the div above.

jonkoops commented on themes/src/main/resources/theme/base/login/resources/js/passwordEntropy.js:
```diff
@@ -0,0 +1,50 @@
+const clamp = (number, min, max) => {
+    return Math.max(min, Math.min(number, max));
+}
```
Can be shorted by using the implicit return.
```suggestion
const clamp = (number, min, max) => Math.max(min, Math.min(number, max));
```

jonkoops commented on themes/src/main/resources/theme/base/login/resources/js/passwordEntropy.js:
```diff
@@ -0,0 +1,50 @@
+}
+
+const checkEntropy = (password) =>  {
+    const entropyContainer = document.getElementById("password-entropy");
```
This can be moved outside of the function, as the DOM should be available and there is no need to re-query the element every time.

jonkoops commented:
Aside from the technical implementation, 'password strength' is a relative term that means different things depending on who you ask. Some users might not even want to have this indicator at all. This likely means that if we were to land this feature it would have to be configurable.

ping @keycloak/core

andreas-blaettlinger commented:
@jonkoops Yes exactly, I agree, looking forward to discussions. The initial state of the PR was just to provide a small demo on the register page as a basis for discussion; not very nice code-wise but functional. Fixed your remarks anyway :)

pedroigor commented:
@andreas-blaettlinger It is a very nice start to start thinking about something more generic.

@jonkoops I would say also configurable in the sense that it should take into account the password policies set to the realm.

jbman commented:
@pedroigor A strength indicator is usually used independently of password rules. [Patternfly](https://www.patternfly.org/components/password-strength/design-guidelines/) has a nice definition: "A password strength indicator is displayed to a user after they have entered a password in a field. It allows the user to understand the strength of a password **after** it has met the password requirements."

I agree with configurability if it should be shown or not. But which mechanism to use? For our custom themes we use localized messages fro configurability, e.g. "passwordStrengthIndicator.show" = false.

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -104,6 +104,17 @@
                     </div>
+                    <div class="${properties.kcInputWrapperClass!}" id="password-entropy" data-password-input="password">
+                        <div class="progress">
+                            <div class="progress-bar" role="status" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100"
```
The `status` role doesn't understand the `value*` attributes, so they can be removed. As long as there is a human-readable text that is inside of the element it will be legible to screen readers.
```suggestion
                            <div class="progress-bar" role="status"
```

jonkoops commented on themes/src/main/resources/theme/keycloak/login/resources/css/password-entropy.css:
```diff
@@ -0,0 +1,8 @@
+    display: none;
+}
+
+#password-entropy > .progress {
```
There should be no need to use a child combinator here, using the class name should be more than specific enough.
```suggestion
#password-entropy .progress {
```

jonkoops commented on themes/src/main/resources/theme/keycloak/login/theme.properties:
```diff
@@ -1,7 +1,7 @@
 import=common/keycloak
 
-styles=css/login.css
+styles=css/login.css css/password-entropy.css
```
Perhaps we can add the styles to `login.css` instead? That saves us from needing to request another file.
```suggestion
styles=css/login.css
```

jonkoops commented on themes/src/main/resources/theme/base/login/resources/js/passwordEntropy.js:
```diff
@@ -0,0 +1,42 @@
+    entropyElement.style.display = "block";
+}
+
+const entropyElement = document.querySelector('#password-entropy');
```
Rather than using the `id` of the element I think we should be [doing the same](https://github.com/keycloak/keycloak/blob/ca2c5f688f03c0674eafd3368b93ab0762ad9b86/themes/src/main/resources/theme/base/login/resources/js/passwordVisibility.js#L14-L15) as we do for password visibility. Which is to use the `data` attribute to find the element on the page, and then bootstrapping all of the logic on that element.

Alternatively, since we're already using modern JavaScript, we could go the extra mile and just splurge on making this a full-fledged [Web Component](https://developer.mozilla.org/en-US/docs/Web/API/Web_components).

Thinking about it a bit more, I think Web Components might be a lot better of a harness than the sprinkles of JavaScript we've been doing. Also from a maintainability perspective.

WDYT?

andreas-blaettlinger commented:
@jonkoops Provided a new WebComponent version. Also made some other changes, e.g. I removed the `data-password-input="password"`, instead I search for a `input[type="password"]` within the parent. Let me know what you think about that.

andreas-blaettlinger commented:
There was this hint from @pedroigor to make this feature configurable. Is there any plan how to handle those fine-grained theme settings and where to persist those yet?

thomasdarimont commented:
How about creating a property `showPasswordQuality`   in `themes.properties`?

In register.ftl you could then write:
````
<#if properties.showPasswordQuality == "true">
                      <!-- insert passwort quality js here -->
                        <password-entropy data-password-poor="${msg('poorPassword')}"
                                          data-password-weak="${msg('weakPassword')}"
                                          data-password-good="${msg('goodPassword')}"
                                          data-password-strong="${msg('strongPassword')}">
                        </password-entropy>
</#if>
````
in https://github.com/keycloak/keycloak/blob/main/themes/src/main/resources/theme/base/login/register.ftl#L3
You could then also move the loading of the java script into the if block.

andreas-blaettlinger commented:
Updated PR which now includes the configuration via `theme.properties` as suggested by @thomasdarimont.

Also added the Password Strength Indicator to all relevant FTL pages.

Any feedback welcome. :)

jonkoops commented:
I have no objections against this feature, but it will need to gather some approval from the core team.

ssilvert commented:
I converted this PR to a draft as it looks like that was the intention all along.  And I think this probably needs more discussion.

I agree that the strength indicator should be on by default.  But I don't think existing installations will necessarily like having this appear when they upgrade.  So we need to decide which way to go on that.

Also, it looks like this doesn't satisfy PatternFly's recommendation that the indicator only comes on after the general password requirements are met?  And that leads us into a discussion of the oft-requested feature of displaying those password requirements to the user.  The user should be told what the requirements are and also should be told how to create a stronger password.  I'm not saying we have to implement that other feature to merge this PR, but the two features are closely related.

Lastly, we will need some documentation on how to customize passwordEntropy.js.

jbman commented:
Hi @ssilvert, 

the intention is also to get a strength indicator into Keycloak to fulfill [ASVS v4.0 Password recommendation 2.1.8](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x11-V2-Authentication.md#v21-password-security-requirements):
> 2.1.8 | Verify that a password strength meter is provided to help users set a stronger password.

On by-default would be good, on by default at new realms and not at existing ones would be best.

But it would also be fine if it could be activated like the password policies. A policy "Show strength indicator" could be added. It wouldn't be rule which prevents any passwords, but password settings would be in one place.

![image](https://github.com/keycloak/keycloak/assets/789496/08a08d91-23f3-4da7-b532-f7217ee5cb5c)

The Patternfly recommendation that "strength indicator is displayed after the password has met all password requirements." could be realized, as soon as the Patternfly component is used and password rules are also displayed to the user beforehand.
In a setup with only one minimum length rule, the strength-indicator is still helpful to encourage strong passwords even if the minimum length is displayed as error after sending the form.

MarcelSz commented:
This idea is nice. However, I'm thinking, how to return password policy from KeyCloak to js. I have many realms with different policies.

As you know, password policies are stored as a single string: "length(8) and maxLength(64) and upperCase(1) and lowerCase(1)". I'd like to have a possibility to validate password in real time - instant feedback. Maybe you have any hints, how to start?
